### PR TITLE
core: make main tsc compile cacheable

### PR DESCRIPTION
**Pull request by brendankenny**
The main typescript compilation (from the root `tsconfig.json`) hasn't had `emitDeclarationOnly` turned on because of compilation errors I ran into in #12914 when I initially enabled that. Not having declaration emit was ok, though, because it's a compilation root, not a dependency for any of our other tsconfigs, so it didn't need a `d.ts` emit for anything. The downside is that less of it is cacheable and more has to be re-done on each compile, even if nothing in the core files has changed.

I ran into this problem again in a different context and it turns out it's fairly simple, only a single error (I thought it was the first of many), and it even tells you exactly what you need to do to fix it.

> `Declaration emit for this file requires using private name 'NetworkRecords' from module '"lighthouse/lighthouse-core/computed/network-records"'. An explicit type annotation may unblock declaration emit.`

This allows fully caching the core compile, so if e.g. you're working on report stuff, you'll only spend ~100ms instead of 5s recompiling core files.

### CODE REVIEW

File: lighthouse-core/index.js [hunk 1]
```diff
@@ -70,6 +70,10 @@ lighthouse.getAuditList = Runner.getAuditList;
 lighthouse.traceCategories = require('./gather/driver.js').traceCategories;
 lighthouse.Audit = require('./audits/audit.js');
 lighthouse.Gatherer = require('./gather/gatherers/gatherer.js');
+
+// Explicit type reference (hidden by makeComputedArtifact) for d.ts export.
+// TODO(esmodules): should be a workaround for module.export and can be removed when in esm.
+/** @type {typeof import('./computed/network-records.js')} */
```
Comment (brendankenny) [inline lighthouse-core/index.js:76]: kind of a dumb workaround :) The need for it is _kind of_ covered by https://github.com/microsoft/TypeScript/issues/37832#issuecomment-611063969 and https://github.com/microsoft/TypeScript/issues/41672, but it's obviously not fixed and it's not a typedef causing the issue.

tl;dr: typescript creates an (ESM-ish) `export` object for the d.ts file, but needs a type for `NetworkRecords`. It then gets confused by the type because `NetworkRecords` is a computed artifact, which subsumes the original `NetworkRecords` type, but the original `NetworkRecords` is still needed because the computed artifact return type is `NetworkRecords & {request(artifacts: DevTools, cache: LH.Artifacts.ComputedContext): Promise<whatever>}`, hence the `"requires using private name 'NetworkRecords'"` complaint.

"Kind of dumb" because the annotation is exactly the expected type for the `require()`, but just need to explicitly type it to unconfuse the compiler and it should go aways with the conversion to es modules anyways, so 🤷

**Review by connorjclark — APPROVED** [review]
